### PR TITLE
[TRITON]: Mha fused backward kernels

### PR DESCRIPTION
**Pull request by juuso-oskari**
Authors: @juuso-oskari @Chi-Chu319 

This PR adds 2 fused backward kernel implementations for multihead-attention (aiter/ops/triton/mha.py):

_flash_attn_fused_backward:
This backward implementation combines the dk, dv and dq computations into one kernel (_bwd_kernel_dkdvdq_causal) using atomics to do shared update of the key, value and query gradients between workgroups. This way we avoid the cost of recomputing values for dq, which is done in the baseline two kernel solution (_bwd_kernel_dkdv_causal & _bwd_kernel_dq_causal). This costs us however the atomic overhead, which we've tried to mitigate using smart remapping of wid -> pid (less contending pids for concurrent workgroups).

_flash_attn_onekernel_backward:
This backward implementation takes advantage of the fact that _bwd_kernel_dkdv_causal & _bwd_kernel_dq_causal share a similar grid, and combines them to be launched as one kernel with one grid.

Initial bench results:

image

image

### CODE REVIEW

File: aiter/ops/triton/utils/mha_onekernel_utils.py [hunk 1]
```diff
@@ -0,0 +1,1157 @@
...
+            qkv = torch.stack([q, k, v], dim=2)
+        else:
+            raise ValueError(f"Unknown layout: {layout}")
+
+        if is_fp8_dtype:
+            raise ValueError("FP8 not supported qkv packing yet")
+        else:
+            return qkv, do, metadata
+    else:
+        assert False, f"Unsupported packing mode: {packing}"
```
Comment (rahulbatra85) [inline aiter/ops/triton/utils/mha_onekernel_utils.py]: These input helpers should move to op_tests. Also, we already have some for mha, so please make sure they are not repeated

File: aiter/ops/triton/mha.py [hunk 2]
```diff
@@ -2387,181 +2779,1132 @@ def _bwd_kernel_dq_noncausal(
...
+        )
+
+    return dk, dv
+
+
+# @triton.autotune(
+#     configs=causal_autotune_configs,
+#     key=causal_autotune_keys,
+#     use_cuda_graph=True,
+# )
```
Comment (rahulbatra85) [inline aiter/ops/triton/mha.py]: Please remove this

File: aiter/ops/triton/mha.py [hunk 3]
```diff
@@ -3,6 +3,457 @@
...
+        return (
+            (preprocess_autotune_configs, preprocess_autotune_keys),
+            (causal_autotune_configs, causal_autotune_keys),
+            (noncausal_autotune_configs, noncausal_autotune_keys),
+        )
+
+
+(
+    (preprocess_autotune_configs, preprocess_autotune_keys),
+    (causal_autotune_configs, causal_autotune_keys),
```
Comment (rahulbatra85) [inline aiter/ops/triton/mha.py]: Remove these. Can you please move this to JSON file?
I have prototype of tuning script which can take such configs(from a JSON file) and programmatically apply the Triton autotuner decorator. It's working for GEMMs right now. I plan to expand it for other ops.

File: aiter/ops/triton/mha.py [hunk 4]
```diff
@@ -2387,181 +2779,1132 @@ def _bwd_kernel_dq_noncausal(
...
+    tl.atomic_add(dv_ptr + offs_dkdv, dv, mask=mask_kv, sem="relaxed")
+    dk *= sm_scale
+    tl.atomic_add(dk_ptr + offs_dkdv, dk, mask=mask_kv, sem="relaxed")
+
+
+# @triton.autotune(
+#     configs=noncausal_autotune_configs,
+#     key=causal_autotune_keys,
+#     use_cuda_graph=True,
+# )
```
Comment (rahulbatra85) [inline aiter/ops/triton/mha.py]: Remove

**Review by rahulbatra85 — CHANGES_REQUESTED** [review]

**Comment by rahulbatra85** [comment]
Address the failing black CI. Thanks!

File: aiter/ops/triton/mha_bwd_onekernel.py
Comment (rahulbatra85) [inline aiter/ops/triton/mha_bwd_onekernel.py:1]: Please add license to new file like in this PR for existing file. Thanks!
https://github.com/ROCm/aiter/pull/485

File: aiter/ops/triton/mha_fused_bwd.py
Comment (rahulbatra85) [inline aiter/ops/triton/mha_fused_bwd.py:1]: Please add license to new file like in this PR for existing file. Thanks!
https://github.com/ROCm/aiter/pull/485

File: aiter/ops/triton/utils/mha_kernel_utils.py
Comment (rahulbatra85) [inline aiter/ops/triton/utils/mha_kernel_utils.py:1]: Please add license to new file like in this PR for existing file. Thanks!
https://github.com/ROCm/aiter/pull/485

File: op_tests/triton_tests/test_mha_fused_bwd.py
Comment (rahulbatra85) [inline op_tests/triton_tests/test_mha_fused_bwd.py:1]: Please add license to new file like in this PR for existing file. Thanks!
https://github.com/ROCm/aiter/pull/485

File: op_tests/triton_tests/test_mha_onekernel_bwd.py
Comment (rahulbatra85) [inline op_tests/triton_tests/test_mha_onekernel_bwd.py:1]: Please add license to new file like in this PR for existing file. Thanks!
https://github.com/ROCm/aiter/pull/485

File: aiter/ops/triton/mha.py [hunk 5]
```diff
@@ -2783,27 +2739,74 @@ def backward(ctx, do, *args):
...
+                None,
+                None,
+                max_seqlen_q=q.shape[1],
+                max_seqlen_k=k.shape[1],
+                dropout_p=ctx.dropout_p,
+                philox_seed=ctx.philox_seed,
+                philox_offset=ctx.philox_offset,
+            )
+        else:
+            _flash_attn_backward(
```
Comment (rahulbatra85) [inline aiter/ops/triton/mha.py]: Please remove the old backward kernels. Only keep the two new ones

File: op_tests/triton_tests/test_mha_fused_bwd.py
Comment (rahulbatra85) [inline op_tests/triton_tests/test_mha_fused_bwd.py:1]: Please merge these tests into the exiting test_mha.py

File: op_tests/triton_tests/test_mha_onekernel_bwd.py
Comment (rahulbatra85) [inline op_tests/triton_tests/test_mha_onekernel_bwd.py:1]: Please merge these tests into the exiting test_mha.py

**Review by rahulbatra85 — CHANGES_REQUESTED** [review]
Remove all code related(op, test and benchmark) to the old backward kernel and keep only the two new ones

File: aiter/ops/triton/mha.py [hunk 6]
```diff
@@ -3048,6 +3113,8 @@ def flash_attn_fp8_func(
     deterministic=False,
     return_lse=False,
     return_attn_probs=False,
+    fused_backward=False,
+    onekernel_backward=False,
```
Comment (rahulbatra85) [inline aiter/ops/triton/mha.py]: Remove these args from the top level API. Instead add a setter function which sets a global variable to control which kernel to use. See the MoE code for example https://github.com/ROCm/aiter/blob/main/aiter/ops/triton/moe_op.py#L18

File: aiter/ops/triton/mha.py [hunk 7]
```diff
@@ -2902,6 +2908,8 @@ def flash_attn_func(
         return_lse,
         return_attn_probs,
         torch.is_grad_enabled(),
+        fused_backward,
+        onekernel_backward,
```
Comment (rahulbatra85) [inline aiter/ops/triton/mha.py]: Remove these args from the top level API. Instead add a setter function which sets a global variable to control which kernel to use. See the MoE code for example https://github.com/ROCm/aiter/blob/main/aiter/ops/triton/moe_op.py#L18

File: aiter/ops/triton/mha.py [hunk 8]
```diff
@@ -3207,6 +3328,8 @@ def flash_attn_varlen_func(
     return_lse=False,
     return_attn_probs=False,
     block_table=None,
+    fused_backward=False,
+    onekernel_backward=False,
```
Comment (rahulbatra85) [inline aiter/ops/triton/mha.py]: Remove these args from the top level API. Instead add a setter function which sets a global variable to control which kernel to use. See the MoE code for example https://github.com/ROCm/aiter/blob/main/aiter/ops/triton/moe_op.py#L18

File: aiter/ops/triton/mha.py [hunk 9]
```diff
@@ -3450,6 +3632,8 @@ def flash_attn_varlen_fp8_func(
...
```
Comment (rahulbatra85) [inline aiter/ops/triton/mha.py]: Remove these args from the top level API. Instead add a setter function which sets a global variable to control which kernel to use. See the MoE code for example https://github.com/ROCm/aiter/blob/main/aiter/ops/triton/moe_op.py#L18
